src/data_morph/morpher.py

```diff
@@ -471,20 +483,27 @@ def _tweening(
             max_value=max_shake,
         )
 
+        x, y = (
+            start_shape.df['x'].to_numpy(copy=True),
+            start_shape.df['y'].to_numpy(copy=True),
+        )
+
         for i in self._looper(
             iterations, leave=True, ascii=True, desc=f'{target_shape} pattern'
         ):
             perturbed_data = self._perturb(
-                morphed_data.copy(),
+                np.copy(x),
+                np.copy(y),
                 target_shape=target_shape,
                 shake=get_current_shake(i),
                 allowed_dist=allowed_dist,
                 temp=get_current_temp(i),
                 bounds=start_shape.morph_bounds,
             )
 
-            if self._is_close_enough(start_shape.df, perturbed_data):
-                morphed_data = perturbed_data
+            if self._is_close_enough(x, y, *perturbed_data):
+                x, y = perturbed_data
+                morphed_data = pd.DataFrame({'x': x, 'y': y})
```

Comment:
This isn't necessary in the loop with switch to NumPy. We can have `_record_frames()` only make the DataFrame if we need to save the CSV. The `plot()` function can be reworked to use NumPy, and to return the DataFrame at the end of this method, we can do that outside of this loop instead of doing it thousands of times.